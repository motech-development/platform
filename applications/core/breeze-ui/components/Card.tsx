import { FC, ReactNode } from 'react';

type TPadding = 'none' | 'sm' | 'md' | 'lg';

const styles = (padding: TPadding, flex: boolean) => {
  const flexStyles = flex ? 'flex flex-col items-center' : '';

  let paddingStyles: string;

  switch (padding) {
    case 'sm':
      break;
    case 'md':
      break;
    case 'lg':
      paddingStyles = 'p-5';
      break;
    default:
  }

  return `${paddingStyles} ${flexStyles}`;
};

const Card: FC<{
  children: ReactNode;
  className?: string;
  flex?: boolean;
  padding?: TPadding;
}> = ({ children, className, flex = false, padding = 'md' }) => (
  <div
    className={`bg-gray-100 border-b-2 border-gray-300 text-gray-800 shadow-lg ${className} ${styles(
      padding,
      flex,
    )}`}
  >
    {children}
  </div>
);

export default Card;
