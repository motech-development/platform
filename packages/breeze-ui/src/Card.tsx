import React, { FC } from 'react';
import styled from 'styled-components';

const BaseCard = styled.div``;

const Card: FC = ({ children }) => {
  return <BaseCard>{children}</BaseCard>;
};

export default Card;
