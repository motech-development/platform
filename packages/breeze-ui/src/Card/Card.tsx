import React, { FC, ReactNode } from 'react';
import styled, { ThemeProvider } from 'styled-components';

type CardPadding = 'sm' | 'md' | 'lg';

interface ICardTheme {
  [name: string]: {
    padding: string;
  };
}

const cardTheme: ICardTheme = {
  lg: {
    padding: '20px',
  },
  md: {
    padding: '10px',
  },
  sm: {
    padding: '0',
  },
};

interface IBaseCard {
  padding: CardPadding;
}

const BaseCard = styled.div<IBaseCard>`
  ${({ theme, padding }) => `
    background: #fff;
    padding: ${theme[padding].padding}
  `}
`;

export interface ICardProps {
  children: ReactNode;
  padding?: CardPadding;
}

const Card: FC<ICardProps> = ({ children, padding = 'sm' }) => {
  return (
    <ThemeProvider theme={cardTheme}>
      <BaseCard padding={padding}>{children}</BaseCard>
    </ThemeProvider>
  );
};

export default Card;
