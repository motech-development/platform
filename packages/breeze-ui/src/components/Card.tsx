import { FC, ReactNode } from 'react';
import { classNames, spacingClass, themeClass } from '../utils/className';
import TSpacing from '../utils/spacing';
import TTheme from '../utils/theme';

export interface ICardProps {
  children: ReactNode;
  className?: string;
  flex?: boolean;
  padding?: TSpacing;
  theme?: TTheme;
}

const Card: FC<ICardProps> = ({
  children,
  className = '',
  flex = false,
  padding = 'md',
  theme = 'secondary',
}) => (
  <>
    {/* @tailwind: bg-blue-100 border-blue-300 text-blue-800 */}
    {/* @tailwind: bg-gray-100 border-gray-300 text-gray-800 */}
    {/* @tailwind: bg-green-100 border-green-300 text-green-800 */}
    {/* @tailwind: bg-red-100 border-red-300 text-red-800 */}
    {/* @tailwind: bg-yellow-100 border-yellow-300 text-yellow-800 */}
    {/* @tailwind: p-0 p-2 p-4 p-6 */}
    <div
      className={classNames(
        'border-b-2 shadow-lg',
        className,
        flex ? 'flex flex-col items-center' : '',
        spacingClass(padding, 'p-{spacing}'),
        themeClass(theme, 'bg-{theme}-100 border-{theme}-300 text-{theme}-800'),
      )}
    >
      {children}
    </div>
  </>
);

export default Card;
