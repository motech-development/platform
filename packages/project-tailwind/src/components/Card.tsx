import { ComponentPropsWithoutRef, forwardRef } from 'react';
import {
  Sizing,
  Themes,
  TSizing,
  TTheme,
  useTailwind,
} from '../utilities/tailwind';

/** Card component props */
export interface ICardProps extends ComponentPropsWithoutRef<'div'> {
  /** Apply flexbox styles */
  flex?: boolean;

  /** Component padding */
  padding?: TSizing;

  /** Componenet theme */
  theme?: TTheme;
}

/**
 * Flexible and extensible content container
 *
 * @param props - Component props
 *
 * @returns Card componenet
 */
export const Card = forwardRef<HTMLDivElement, ICardProps>(
  (
    {
      className,
      flex = false,
      padding = Sizing.MD,
      theme = Themes.SECONDARY,
      ...rest
    },
    ref,
  ) => {
    const { createStyles } = useTailwind(theme, padding);

    const cardStyles = createStyles({
      classNames: [
        'border-b-2 shadow-lg',
        {
          'flex flex-col items-center': flex,
        },
        className,
      ],
      sizing: {
        lg: ['p-6'],
        md: ['p-4'],
        none: ['p-0'],
        sm: ['p-2'],
      },
      theme: {
        danger: ['bg-red-100 border-red-300 text-red-800'],
        primary: ['bg-blue-100 border-blue-300 text-blue-800'],
        secondary: ['bg-gray-100 border-gray-300 text-gray-800'],
        success: ['bg-green-100 border-green-300 text-green-800'],
        warning: ['bg-yellow-100 border-yellow-300 text-yellow-800'],
      },
    });

    return <div className={cardStyles} ref={ref} {...rest} />;
  },
);
