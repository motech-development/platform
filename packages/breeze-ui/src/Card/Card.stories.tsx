import { select, text, withKnobs } from '@storybook/addon-knobs';
import BaseStyles from '../BaseStyles/BaseStyles';
import Card from './Card';

export default {
  component: Card,
  decorators: [withKnobs],
};

const padding = {
  Large: 'lg',
  Medium: 'md',
  Small: 'sm',
};

export const BasicCard = {
  name: 'Basic card',
  render: () => (
    <>
      <BaseStyles />

      <Card padding={select('Padding', padding, 'md') as 'sm' | 'md' | 'lg'}>
        {text('Text', 'Hello world')}
      </Card>
    </>
  ),
};
