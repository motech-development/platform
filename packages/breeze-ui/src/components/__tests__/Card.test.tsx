import { render } from '@testing-library/react';
import Card from '../Card';

describe('Card', () => {
  describe('padding', () => {
    it('should have the correct padding when set to "sm"', () => {
      const { container } = render(
        <Card padding="sm">
          <p data-testid="content">Hello world</p>
        </Card>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct padding when set to "md"', () => {
      const { container } = render(
        <Card padding="md">
          <p data-testid="content">Hello world</p>
        </Card>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct padding when set to "lg"', () => {
      const { container } = render(
        <Card padding="lg">
          <p data-testid="content">Hello world</p>
        </Card>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct padding when set to "none"', () => {
      const { container } = render(
        <Card padding="none">
          <p data-testid="content">Hello world</p>
        </Card>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('theme', () => {
    it('should set to "secondary" by default', () => {
      const { container } = render(
        <Card>
          <p data-testid="content">Hello world</p>
        </Card>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct theme when set to "primary"', () => {
      const { container } = render(
        <Card theme="primary">
          <p data-testid="content">Hello world</p>
        </Card>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct theme when set to "secondary"', () => {
      const { container } = render(
        <Card theme="secondary">
          <p data-testid="content">Hello world</p>
        </Card>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct theme when set to "danger"', () => {
      const { container } = render(
        <Card theme="danger">
          <p data-testid="content">Hello world</p>
        </Card>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct theme when set to "success"', () => {
      const { container } = render(
        <Card theme="success">
          <p data-testid="content">Hello world</p>
        </Card>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct theme when set to "warning"', () => {
      const { container } = render(
        <Card theme="warning">
          <p data-testid="content">Hello world</p>
        </Card>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  it('should output child', async () => {
    const { findByTestId } = render(
      <Card>
        <p data-testid="content">Hello world</p>
      </Card>,
    );

    await expect(findByTestId('content')).resolves.toBeInTheDocument();
  });

  it('should set correct custom classes', () => {
    const { container } = render(
      <Card className="my-custom-class">
        <p data-testid="content">Hello world</p>
      </Card>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should set flexbox classes', () => {
    const { container } = render(
      <Card flex>
        <p data-testid="content">Hello world</p>
      </Card>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
