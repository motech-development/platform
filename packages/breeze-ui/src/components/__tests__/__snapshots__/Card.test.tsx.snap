// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card padding should have the correct padding when set to "lg" 1`] = `
<div
  class="border-b-2 shadow-lg p-6 bg-gray-100 border-gray-300 text-gray-800"
>
  <p
    data-testid="content"
  >
    Hello world
  </p>
</div>
`;

exports[`Card padding should have the correct padding when set to "md" 1`] = `
<div
  class="border-b-2 shadow-lg p-4 bg-gray-100 border-gray-300 text-gray-800"
>
  <p
    data-testid="content"
  >
    Hello world
  </p>
</div>
`;

exports[`Card padding should have the correct padding when set to "none" 1`] = `
<div
  class="border-b-2 shadow-lg p-0 bg-gray-100 border-gray-300 text-gray-800"
>
  <p
    data-testid="content"
  >
    Hello world
  </p>
</div>
`;

exports[`Card padding should have the correct padding when set to "sm" 1`] = `
<div
  class="border-b-2 shadow-lg p-2 bg-gray-100 border-gray-300 text-gray-800"
>
  <p
    data-testid="content"
  >
    Hello world
  </p>
</div>
`;

exports[`Card should set correct custom classes 1`] = `
<div
  class="border-b-2 shadow-lg my-custom-class p-4 bg-gray-100 border-gray-300 text-gray-800"
>
  <p
    data-testid="content"
  >
    Hello world
  </p>
</div>
`;

exports[`Card should set flexbox classes 1`] = `
<div
  class="border-b-2 shadow-lg flex flex-col items-center p-4 bg-gray-100 border-gray-300 text-gray-800"
>
  <p
    data-testid="content"
  >
    Hello world
  </p>
</div>
`;

exports[`Card theme should have the correct theme when set to "danger" 1`] = `
<div
  class="border-b-2 shadow-lg p-4 bg-red-100 border-red-300 text-red-800"
>
  <p
    data-testid="content"
  >
    Hello world
  </p>
</div>
`;

exports[`Card theme should have the correct theme when set to "primary" 1`] = `
<div
  class="border-b-2 shadow-lg p-4 bg-blue-100 border-blue-300 text-blue-800"
>
  <p
    data-testid="content"
  >
    Hello world
  </p>
</div>
`;

exports[`Card theme should have the correct theme when set to "secondary" 1`] = `
<div
  class="border-b-2 shadow-lg p-4 bg-gray-100 border-gray-300 text-gray-800"
>
  <p
    data-testid="content"
  >
    Hello world
  </p>
</div>
`;

exports[`Card theme should have the correct theme when set to "success" 1`] = `
<div
  class="border-b-2 shadow-lg p-4 bg-green-100 border-green-300 text-green-800"
>
  <p
    data-testid="content"
  >
    Hello world
  </p>
</div>
`;

exports[`Card theme should have the correct theme when set to "warning" 1`] = `
<div
  class="border-b-2 shadow-lg p-4 bg-yellow-100 border-yellow-300 text-yellow-800"
>
  <p
    data-testid="content"
  >
    Hello world
  </p>
</div>
`;

exports[`Card theme should set to "secondary" by default 1`] = `
<div
  class="border-b-2 shadow-lg p-4 bg-gray-100 border-gray-300 text-gray-800"
>
  <p
    data-testid="content"
  >
    Hello world
  </p>
</div>
`;
